move starts_with from Utils to DB -- where it is used

hannesm opened the pull request:
simplify and move set_fixed_string into single user in Formats

shrink the Utils module, avoid opening it in Formats and GUI (where it is not
used much).

this is on the road to address #12

talex5 commented on lib/gUI.ml:
```diff
@@ -156,10 +155,10 @@ let connect ~domid () =
   set_gui_protocol_version_version version qubes_gui_protocol_version_linux;
   QV.send qv [version] >>= function
-  | `Eof -> Lwt.fail (error "End-of-file sending protocol version")
+  | `Eof -> Lwt.fail (Utils.error "End-of-file sending protocol version")
```
I think these can just be e.g.

```suggestion
  | `Eof -> Lwt.fail_with "End-of-file sending protocol version"
```

talex5 reviewed (commented):
Looks OK, but I think we could just remove the whole module. e.g.

- `raise (Utils.error` -> `Fmt.failwith`
- `Lwt.fail (Utils.error` -> `Lwt.fail_with`
- `starts_with` -> `Astring.String.is_prefix` (or wait for OCaml 4.12)
- `return`/`fail` -> `Lwt.return`/`Lwt.fail`

hannesm commented:
ok, that was my goal in the end. I'll wait for #57 to be merged, and rebase and revise this PR. I'm not eager to introduce a dependency (on astrng) just for `prefix` -- so I suggest to leave it in here until we drop < 4.12 support.

thanks for your review.

hannesm commented:
I force-pushed to this branch -- the Utils module is gone now.

talex5 commented on lib/dB.ml:
```diff
@@ -1,15 +1,24 @@
-  | `Eof -> fail (Failure "qubesdb-agent: end-of-file from QubesDB!")
-  | `Error (`Unknown msg) -> fail (error "qubesdb-agent: %s" msg)
+  | `Eof -> Lwt.fail_with "qubesdb-agent: end-of-file from QubesDB!"
+  | `Error (`Unknown msg) -> Lwt.fail (Fmt.failwith "qubesdb-agent: %s" msg)
```
`Fmt.failwith` raises an exception, so either remove the unreachable `Lwt.fail` or use e.g. `Fmt.strf` to generate the error.
```suggestion
  | `Error (`Unknown msg) -> Fmt.failwith "qubesdb-agent: %s" msg
```

talex5 commented on lib/dB.ml:
```diff
@@ -58,19 +67,19 @@ let full_db_sync t =
         t.store <- t.store |> KeyMap.add path data;
         loop ()
-    | ty, _, _ -> fail (error "Unexpected QubesDB message: %s" (qdb_msg_to_string ty)) in
+    | ty, _, _ -> Lwt.fail (Fmt.failwith "Unexpected QubesDB message: %s" (qdb_msg_to_string ty)) in
```
Same here.

talex5 commented on lib/rExec.ml:
```diff
@@ -194,14 +193,14 @@ let send_hello t =
-  | `Ok (ty, _) -> fail (error "Expected msg_hello, got %ld" (int_of_type ty))
+  | `Eof -> Lwt.fail_with "End-of-file waiting for msg_hello"
+  | `Ok (`Hello, resp) -> Lwt.return (get_peer_info_version resp)
+  | `Ok (ty, _) -> Lwt.fail (Fmt.failwith "Expected msg_hello, got %ld" (int_of_type ty))
```
Same here.

talex5 commented on lib/rExec.ml:
```diff
@@ -215,15 +214,15 @@ let with_flow ~ty ~domid ~port fn =
         (fun () ->
           recv_hello client >>= function
-          | version when version < 2l -> fail (error "Unsupported qrexec version %ld" version)
+          | version when version < 2l -> Lwt.fail (Fmt.failwith "Unsupported qrexec version %ld" version)
```
Same here.

talex5 commented on lib/rExec.ml:
```diff
@@ -248,12 +247,12 @@ let port_of_int i =
     match cmd |> split ':' with
-    | None -> fail (error "Missing ':' in %S" cmd)
-    | Some (user, cmd) -> return (user, cmd)
+    | None -> Lwt.fail (Fmt.failwith "Missing ':' in %S" cmd)
```
Same here.

talex5 commented on lib/rExec.ml:
```diff
@@ -377,9 +376,9 @@ let connect ~domid () =
   send_hello t.t >>= fun () ->
   recv_hello t.t >>= function
-  | version when version < 2l -> fail (error "Unsupported qrexec version %ld" version)
+  | version when version < 2l -> Lwt.fail (Fmt.failwith "Unsupported qrexec version %ld" version)
```
Same here.

talex5 reviewed (commented):
`Lwt.fail (Fmt.failwith` seems a bit odd, as `failwith` never returns.

hannesm commented:
thanks @talex5, fixed in a subsequent commit. I just dropped the `Lwt.fail` call (due to `Lwt.fail` documentation: Whenever possible, it is recommended to use [raise exn] instead, as [raise] captures a backtrace, while [Lwt.fail] does not. If you call [raise exn] in a callback that is expected by Lwt to return a promise, Lwt will automatically wrap [exn] in a rejected promise, but the backtrace will have been recorded by the OCaml runtime. Use [Lwt.fail] only when you specifically want to create a rejected promise, to pass to another function,  or store in a data structure.)

talex5 commented:
Thanks!
